### PR TITLE
Improve tag replacement

### PR DESCRIPTION
**Pull request by phackstock**
This PR implements an improved tag replacement logic. 
The new logic is:
* Find all occurrences of `"{tag}"`
* If the tag has the same attribute as the base code, replace the value of the attribute
* If the tag does not have the same attribute, replace with the name of the tag. 

Also, since I added the property `Code.flattened_dict` I also made sure to use it in `CodeList.codelist_repr`.

### CODE REVIEW

File: doc/source/user_guide/codelist.rst [hunk 1]
```diff
@@ -50,10 +50,11 @@ to be used as tag. The files defining the tags must be named like ``tag_*.yaml``
      - Some Key:
          description: a short description of the key
 
-When importing a tagged codelist, any occurrence of ``{Tag}`` in the name of a code will
-be replaced by the corresponding element in the Tag dictionary. If the tag does not contain an element with ``{Tag}``, the tag name is used.
+When importing a tagged codelist, code containing ``{Tag}`` in its name of a code will
```
Comment (danielhuppmann) [inline doc/source/user_guide/codelist.rst:53]:
```suggestion
When importing a tagged codelist, any code containing ``{Tag}`` in its name will
```

File: tests/data/tagged_codelist/tag_sector.yaml [hunk 2]
```diff
@@ -6,4 +6,3 @@
       weight: Energy
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/tag_sector.yaml:6]: Are the the "value"-entries actually used anywhere?

**Review by danielhuppmann — APPROVED** [review]
Looks good, just two minor suggestions!

File: tests/data/tagged_codelist/tag_sector.yaml (hunk 2 above)
Comment (phackstock) [reply tests/data/tagged_codelist/tag_sector.yaml:6]: What do you mean by the "value" entries?

Comment (danielhuppmann) [reply tests/data/tagged_codelist/tag_sector.yaml:6]: There are two attributes "value_bool" and "value_number" just above the commented line...

Comment (phackstock) [reply tests/data/tagged_codelist/tag_sector.yaml:6]: Ah right, sorry didn't see those. I don't think they're used, I'll double check. Should I remove them if they're not used?

Comment (phackstock) [reply tests/data/tagged_codelist/tag_sector.yaml:4]: Removed it, didn't break the tests on my machine. If they also run through here I'll go ahead with the merge.

**Pull request merged**